### ISSUE TITLE
Duplicate Type Generation

### ISSUE THREAD

**Issue by felixkmh** [JasperFx/marten#3083]
I ran into a memory issue that ultimately comes from duplicate code generation for document types. Because certain events were processed in parallel in batches of size **n**, when the first batch was processed, code was generated for each type up to **n** times and loaded into the AppDomain, causing the application to consume multiple gigabytes of memory in some cases.

I could fix this behavior by registering the types upfront and setting the code generation mode to `Auto`. But my suggestion would be to make `ProviderGraph.cs` thread-safe such that code generation is guaranteed to only occur once per type.

**Comment by jeremydmiller** [comment]
What version of Marten? We've fought this one several times already

**Comment by felixkmh** [comment]
Oh, sorry for not mentioning that. It first occured on 6.0.4, but also 6.4.1 did not seem to fix this issue.

**Comment by jeremydmiller** [comment]
Okay. All I'm going to do is the exact same thing, so if you're interested, "I take pull requests"

**Comment by felixkmh** [comment]
I'll take you up on that. I would like to check whether it would be feasable to have one lock per type though. If not, I'll just go with the safer option.

**Comment by jeremydmiller** [comment]
Stick the lock if anything in GeneratedProjection.CodeGen, not in ProviderGraph? that's roughly how we did the same kind of locking for Wolverine for similar issues

**Comment by felixkmh** [comment]
My understanding was that multiple threads can walk over `if (_storage.TryFind(documentType, out var stored)) { ... }`  for the same `documentType` here:
```csharp
public DocumentProvider<T> StorageFor<T>() where T : notnull
{
        var documentType = typeof(T);

        if (_storage.TryFind(documentType, out var stored))
        {
            return stored.As<DocumentProvider<T>>();
        }
        // Generate code (possibly time intesive) and then update _storage
}
```
and enter the code generation part (i.e. `builder.InitializeSynchronously(rules, _options, null);`) if they run in parallel, because `_storage` is only updated after the code generation. That's why I considered to lock in there. Am I on the wrong track here?

**Comment by felixkmh** [comment]
I put together a script to illustrate my idea:

<details>
<summary>Script</summary>

```csharp
using System.Collections.Concurrent;
using System.Threading;
using System.Diagnostics;

ConcurrentDictionary<Type, object> _storage = new();
Dictionary<Type, object> typeLockObjects = new();

public object GetTypeLock(Type type)
{
    lock(typeLockObjects)
    {
        // return existing lock object or generate a new one if none exists
        return typeLockObjects.TryGetValue(type, out var typeLock) ? typeLock : typeLockObjects[type] = new();
    }
}

public object GenerateCodeForType<T>()
{
    // Shortcut without locking. We are only reading here, so it should be fine.
    if (_storage.TryGetValue(typeof(T), out var generated))
    {
        return generated;
    }

    lock(GetTypeLock(typeof(T)))
    {
        // Re-evaluate _storage in case it was changed while waiting for the type lock
        if (_storage.TryGetValue(typeof(T), out var generated2))
        {
            return generated2;
        }

        // Generate code for T and update _storage
        Thread.Sleep(500);
        Console.WriteLine($"Generated for type {typeof(T).Name}");
        // writing needs to be synchronized, here by the ConcurrentDictionary
        return _storage[typeof(T)] = new();
    }
}

List<Task> tasks = new();

Stopwatch sw = Stopwatch.StartNew();

for (int i = 0; i < 50; ++i)
{
    tasks.Add(Task.Run(() => GenerateCodeForType<Action>()));
    tasks.Add(Task.Run(() => GenerateCodeForType<Byte>()));
    tasks.Add(Task.Run(() => GenerateCodeForType<Comparer>()));
    tasks.Add(Task.Run(() => GenerateCodeForType<DateTime>()));
    tasks.Add(Task.Run(() => GenerateCodeForType<Encoder>()));
    tasks.Add(Task.Run(() => GenerateCodeForType<Func<float>>()));
    tasks.Add(Task.Run(() => GenerateCodeForType<Guid>()));
    tasks.Add(Task.Run(() => GenerateCodeForType<Hashtable>()));
    tasks.Add(Task.Run(() => GenerateCodeForType<int>()));
}

await Task.WhenAll(tasks);

sw.Stop();

Console.WriteLine($"Completed after {sw.ElapsedMilliseconds}ms")
```

</details>

**Comment by jeremydmiller** [comment]
Meh, I see why it makes sense to put it in `ProviderGraph` because the ICodeFile isn't stateful. I'd vote to just use a single lock object though. I don't think there's enough concurrency like that to matter.

**Comment by jeremydmiller** [comment]
Closed by #3109
